### PR TITLE
XLA Parallel reduce

### PR DESCRIPTION
**Pull request by trentlo**
Extend the XLA codegen to generate parallel reductions when there are multiple
reduce instructions in a fusion computation. This paves a way for horizontal input
fusion in the future.

### Performance Numbers
Now we see ~3% e2e gain for NVIDIA JoC BERT.

For `ManyParallelReductions` with 128 reduce instructions in the unittest, the
execution time is reduced from 325us to 3.9us (83X), reported by nvprof as below.

Before:

            Type  Time(%)      Time     Calls       Avg       Min       Max  Name

                   32.50%  325.54us         1  325.54us  325.54us  325.54us  fusion

After:

            Type  Time(%)      Time     Calls       Avg       Min       Max  Name

                    0.59%  3.9030us         1  3.9030us  3.9030us  3.9030us  fusion

### High-level descriptions about the Algorithm
The overall idea is to divide the reduce instructions in a fusion into groups for parallelization, as implemented by `DivideOutputInstructionsIntoGroups()`. Different groups will be executed in parallel, while reduce instructions in the same group are run sequentially. Generally speaking, we'd like to run the reduce instructions in parallel without incurring too much re-computation overhead. The current heuristic is to place reduce instructions who share nothing or only
 (broadcasted) scalars/constants into different groups; otherwise, they are placed in the same group. Non-reduce instructions always go with the reduce instructions into the same group so long as they share any predecessors.

To avoid complicating the (already complicated) index calculation in reduce code generation, we use raw (CUDA) block_id_y to select the reduce groups (instead of thread_id_x/block_id_x). That is, a block_id_y is assigned to a group and so different groups can be run in parallel. Separation of the reduce codegen index calculation (using raw dim x) and parallelization (using raw dim y) should also provide better future maintainability.

### CODE REVIEW

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc [hunk 1]
```diff
@@ -2991,14 +2992,35 @@ void IrEmitterUnnested::EmitPrintfWithThreadId(
...
+  if (!unnested_hlo.IsMultiOutputFusion()) {
+    return ShapeIndex({});
+  }
+  const auto& all_outputs = unnested_hlo.fused_expression_root()->operands();
+  for (size_t i = 0; i < all_outputs.size(); ++i) {
+    if (all_outputs[i] == &out_instr) {
+      return ShapeIndex({i});
+    }
+  }
+  CHECK(false) << " Fusion root does not contain output instruction; "
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:3010]: LOG(FATAL)

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc [hunk 2]
```diff
@@ -3748,86 +3771,211 @@ ReductionCodegenInfo IrEmitterUnnested::ComputeReductionCodegenInfo(
...
-
   const Shape& input_shape = first_reduce->operand(0)->shape();
   // The layout of a reduction input is either set by LayoutAssignment for
   // unnested kReduce or by InstructionFusion for fused kReduce.
   CHECK(input_shape.has_layout()) << "LayoutAssignment or InstructionFusion "
                                      "doesn't set the input layout of "
                                   << first_reduce->ToString();
 
+  // Group output instructions. Each group will be executed in parallel.
+  auto instr_groups =
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:3935]: Please avoid `auto` here and below (apart from lambdas, where the type is non-obvious)

File: tensorflow/compiler/xla/service/gpu/buffer_comparator.cc [hunk 3]
```diff
@@ -613,10 +613,13 @@ static StatusOr<bool> DeviceCompare(se::Stream* stream,
   LaunchDimensions dim =
       CalculateLaunchDimensions(buffer_shape, gpu_device_info);
 
-  stream->ThenLaunch(se::ThreadDim(dim.threads_per_block()),
-                     se::BlockDim(dim.block_count()), *comparison_kernel,
-                     lhs_typed, rhs_typed, static_cast<float>(kTolerance),
-                     buffer_size, out_param.cref());
+  auto thread_counts = dim.thread_counts_per_block();
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/buffer_comparator.cc:616]: Please avoid `auto` when non-obvious

File: tensorflow/compiler/xla/service/gpu/launch_dimensions.h [hunk 4]
```diff
@@ -29,24 +29,37 @@ namespace gpu {
 // number of threads per block.
 class LaunchDimensions {
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/launch_dimensions.h:30]: Could you provide some more info in the PR description why 3-dimensions are needed for the launch dimensions?

**Review by cheshire — COMMENTED** [review]
Some nitpicks. Overall looks great, thanks!

**Review by cheshire — CHANGES_REQUESTED** [review]
Could you give a high-level description of the algorithm in the PR description? It's a bit hard to follow just from the code.

File: tensorflow/compiler/xla/service/gpu/buffer_comparator.cc (hunk 3 above)
Comment (trentlo) [reply tensorflow/compiler/xla/service/gpu/buffer_comparator.cc]: All are fixed.

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc (hunk 1 above)
Comment (trentlo) [reply tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc]: Revised.

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc (hunk 2 above)
Comment (trentlo) [reply tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc]: Applied when applicable.

File: tensorflow/compiler/xla/service/gpu/launch_dimensions.h (hunk 4 above)
Comment (trentlo) [reply tensorflow/compiler/xla/service/gpu/launch_dimensions.h:30]: I added more info in the PR description and more corresponding comments in the codes. Let me know if it helps.

**Review by trentlo — COMMENTED** [review]
Thanks for the review.

I added more descriptions about the algorithm in the PR and also in the codes. Let me know if they help.

**Comment by cheshire** [comment]
Thanks! I'm fine with this as long as nitpicks are addressed.

**Comment by trentlo** [comment]
> Thanks! I'm fine with this as long as nitpicks are addressed.

@cheshire just to make sure that you know I have updated the codes to address your previous comments.

Would you be the main reviewer or @thomasjoerg will take charge?

**Review by cheshire — DISMISSED** [review]

**Comment by cheshire** [comment]
@trentlo I'm fine with merging this.

**Comment by cheshire** [comment]
(provided our internal benchmarks don't regress)

**Comment by trentlo** [comment]
> @trentlo I'm fine with merging this.

Great! I do not see regressions from the NV-internal benchmarks but we will see.

**Comment by trentlo** [comment]
George, could you launch the CI again? Thanks!

There were build errors due to narrowing type conversion (In MacOS CPU Python3 and Bazel Windows GPU). I fixed them by pushing a new commit. An error in Bazel Windows was not seem to related to the PR. Other tests were clean.

**Comment by cheshire** [comment]
@trentlo Could you fix the following warnings?

```
third_party/tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:3007:26: error: non-constant-expression cannot be narrowed from type 'size_t' (aka 'unsigned long') to 'long long' in initializer list [-Wc++11-narrowing]
      return ShapeIndex({i});
                         ^
third_party/tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:3007:26: note: insert an explicit cast to silence this issue
      return ShapeIndex({i});
                         ^
                         static_cast<long long>( )
third_party/tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:3834:50: error: '&&' within '||' [-Werror,-Wlogical-op-parentheses]
         HloOpcode::kBroadcast == instr.opcode() &&
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
third_party/tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:3834:50: note: place parentheses around the '&&' expression to silence this warning
         HloOpcode::kBroadcast == instr.opcode() &&
                                                 ^
         (
third_party/tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:3957:7: error: ignoring return value of function declared with 'warn_unused_result' attribute [-Werror,-Wunused-result]
      EmitIRForReduction(unnested_hlo, instr_groups[i], &reduction_info,
      ^~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
third_party/tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:3979:56: error: non-constant-expression cannot be narrowed from type 'std::__u::vector<std::__u::vector<xla::HloInstruction *, std::__u::allocator<xla::HloInstruction *>>, std::__u::allocator<std::__u::vector<xla::HloInstruction *, std::__u::allocator<xla::HloInstruction *>>>>::size_type' (aka 'unsigned long') to 'tensorflow::int64' (aka 'long long') in initializer list [-Wc++11-narrowing]
      {/*x=*/mapping_scheme.GetNumberOfBlocks(), /*y=*/instr_groups.size(),
                                                       ^~~~~~~~~~~~~~~~~~~
third_party/tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:3979:56: note: insert an explicit cast to silence this issue
      {/*x=*/mapping_scheme.GetNumberOfBlocks(), /*y=*/instr_groups.size(),
                                                       ^~~~~~~~~~~~~~~~~~~
                                                       static_cast<int64>()

```

**Comment by cheshire** [comment]
(we have a long-standing bug to enable these warnings in OSS, but so far it hasn't been done =/)

**Comment by trentlo** [comment]
> @trentlo Could you fix the following warnings?
> 
> ```
> third_party/tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:3007:26: error: non-constant-expression cannot be narrowed from type 'size_t' (aka 'unsigned long') to 'long long' in initializer list [-Wc++11-narrowing]
>       return ShapeIndex({i});
>                          ^
> third_party/tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:3007:26: note: insert an explicit cast to silence this issue
>       return ShapeIndex({i});
>                          ^
>                          static_cast<long long>( )
> third_party/tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:3834:50: error: '&&' within '||' [-Werror,-Wlogical-op-parentheses]
>          HloOpcode::kBroadcast == instr.opcode() &&
>          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
> third_party/tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:3834:50: note: place parentheses around the '&&' expression to silence this warning
>          HloOpcode::kBroadcast == instr.opcode() &&
>                                                  ^
>          (
> third_party/tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:3957:7: error: ignoring return value of function declared with 'warn_unused_result' attribute [-Werror,-Wunused-result]
>       EmitIRForReduction(unnested_hlo, instr_groups[i], &reduction_info,
>       ^~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> third_party/tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:3979:56: error: non-constant-expression cannot be narrowed from type 'std::__u::vector<std::__u::vector<xla::HloInstruction *, std::__u::allocator<xla::HloInstruction *>>, std::__u::allocator<std::__u::vector<xla::HloInstruction *, std::__u::allocator<xla::HloInstruction *>>>>::size_type' (aka 'unsigned long') to 'tensorflow::int64' (aka 'long long') in initializer list [-Wc++11-narrowing]
>       {/*x=*/mapping_scheme.GetNumberOfBlocks(), /*y=*/instr_groups.size(),
>                                                        ^~~~~~~~~~~~~~~~~~~
> third_party/tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:3979:56: note: insert an explicit cast to silence this issue
>       {/*x=*/mapping_scheme.GetNumberOfBlocks(), /*y=*/instr_groups.size(),
>                                                        ^~~~~~~~~~~~~~~~~~~
>                                                        static_cast<int64>()
> ```

PR updated. Please help to launch the CI again. Thanks~

**Review by cheshire — APPROVED** [review]

**Comment by sanjoy** [comment]
CC @timshen91
